src/preprocessing/passes/ackermann.cpp

```diff
@@ -187,9 +188,124 @@ void collectFunctionsAndLemmas(FunctionToArgsMap& fun_to_args,
 
 /* -------------------------------------------------------------------------- */
 
+bool needsReplace(TNode term)
+{
+  if (term.getType().isSort() && !(term.getKind() == kind::BOUND_VARIABLE))
+    return true;
+  return false;
+}
+
+/* Given the lowest capacity requirements for each uninterpreted sorts, assign
+ * a sufficient bit-vector size. Get the converting map */
+void collectUSortsToBV(USortToBVSizeMap& usortCardinality,
+                       vector<TNode>& vec,
+                       SubstitutionMap& sortsToSkolem)
+{
+  NodeManager* nm = NodeManager::currentNM();
+
+  for (TNode term : vec)
+  {
+    if (needsReplace(term))
+    {
+      TypeNode type = term.getType();
+      size_t size = usortCardinality[type].second;
+      if (size == 0)
+      {
+        size = log2(usortCardinality[type].first) + 1;
+        usortCardinality[type].second = size;
+      }
+      Node skolem = nm->mkSkolem(
+          "BVSKOLEM$$",
+          nm->mkBitVectorType(size),
+          "a variable created by the ackermannization "
+          "preprocessing pass, representing term with uninterpreted sorts.");
+      sortsToSkolem.addSubstitution(term, skolem);
+    }
+  }
+}
+
+/* This function returns the list of terms in formula represented by assertions.
+ * We use a BFS to get all terms without duplications. */
+std::vector<TNode> getListOfTerms(AssertionPipeline* assertions)
+{
+  TNodeSet seen;
+  std::vector<TNode> res;
+  for (Node& a : assertions->ref())
+  {
+    if (seen.find(a) == seen.end())
+    {
+      res.push_back(a);
+      seen.insert(a);
+    }
+  }
+  size_t i = 0;
+  while (i < res.size())
+  {
+    for (TNode a : res[i])
+    {
+      if (seen.find(a) == seen.end())
+      {
+        res.push_back(a);
+        seen.insert(a);
+      }
+    }
+    ++i;
+  }
+  return res;
+}
+
+/* This is the top level of converting uninterpreted sorts to bit vectors.
+ * We use BFS to get all terms without duplications, and count the number of
+ * different terms for each uninterpreted sort. Then for each sort, we will
+ * assign a new bit-vector type with a sufficient size.
+ * The size is calculated to have enough capacity, that can accommodate the
+ * variables occured in the original formula. */
+void usortsToBitVectors(const LogicInfo& d_logic,
+                        AssertionPipeline* assertions,
+                        USortToBVSizeMap& usortCardinality,
+                        SubstitutionMap& sortsToSkolem)
+{
+  std::vector<TNode> toProcess = getListOfTerms(assertions);
```

Comment:
As far as I can tell, we later only care about terms in `toProcess` that `needsReplace()`. If that is the case, we shouldn't add them to `toProcess` in the first place (they can already be filtered out in `getListOfTerms()`. This will save time and memory.